smk/gene_abundance.smk

```diff
@@ -378,35 +392,54 @@ rule genome_mapping_profiling:
-                msamtools filter -S -b -l {params.length} -p {wildcards.identity} -z 80 --besthit - > aligned.bam) >& {output.bwa_log}
-        total_reads="$(grep Processed {output.bwa_log} | perl -ne 'm/Processed (\\d+) reads/; $sum+=$1; END{{printf "%d\\n", $sum/2;}}')"
-        #echo $total_reads
-        common_args="--label {wildcards.omics}.{params.sample_alias} --total=$total_reads --mincount={params.mapped_reads_threshold} --pandas"
-        msamtools profile aligned.bam $common_args -o {output.raw_all_seq}     --multi=all  --unit=abund --nolen
-        msamtools profile aligned.bam $common_args -o {output.raw_prop_seq}    --multi=prop --unit=abund --nolen
-        msamtools profile aligned.bam $common_args -o {output.raw_prop_genome} --multi=prop --unit=abund --nolen --genome {input.genome_def}
-        msamtools profile aligned.bam $common_args -o {output.rel_prop_genome} --multi=prop --unit=rel           --genome {input.genome_def}
-
-        samtools sort aligned.bam -o sorted.bam -@ {params.sort_threads} -m {params.sort_memory}G --output-fmt=BAM
-        samtools index sorted.bam sorted.bam.bai -@ {threads}
-
-        rsync -a sorted.bam {output.sorted}
-        rsync -a sorted.bam.bai {output.index}
+                    msamtools filter -S -b -l {params.length} -p {wildcards.identity} -z 80 --besthit - > aligned.bam) >& {output.bwa_log}
+            total_reads="$(grep Processed {output.bwa_log} | perl -ne 'm/Processed (\\d+) reads/; $sum+=$1; END{{printf "%d\\n", $sum/2;}}')"
+            #echo $total_reads
+
+            # Run multiple msamtools modes + samtools sort in parallel using GNU parallel.
+            # We need 4 threads for msamtools and 'params.sort_threads' threads for samtools sort.
+            # We cap it by 'threads' limit to parallel.
+            common_args="--label {wildcards.omics}.{params.sample_alias} --total=$total_reads --mincount={params.mapped_reads_threshold} --pandas"
+            parallel --jobs {threads} <<__EOM__
+msamtools profile aligned.bam $common_args -o {output.raw_all_seq}     --multi=all  --unit=abund --nolen
+msamtools profile aligned.bam $common_args -o {output.raw_prop_seq}    --multi=prop --unit=abund --nolen
+msamtools profile aligned.bam $common_args -o {output.raw_prop_genome} --multi=prop --unit=abund --nolen --genome {input.genome_def}
+msamtools profile aligned.bam $common_args -o {output.rel_prop_genome} --multi=prop --unit=rel           --genome {input.genome_def}
+samtools sort aligned.bam -o sorted.bam -@ {params.sort_threads} -m {params.sort_memory}G --output-fmt=BAM
+__EOM__
+
+            # Index sorted.bam file, while also exporting it
+            parallel --jobs {threads} <<__EOM__
+samtools index sorted.bam sorted.bam.bai -@ {threads}
+rsync -a sorted.bam {output.sorted}
+__EOM__
+
+            # Use bedtools multicov to generate read-counts per gene
+            rsync -a {input.bed_mini} in.bed
```

Comment:
Why not put the mini_bed rsync into the parallel block above?

Reply:
Yes, good point.